Add TT implementation.

DomiKoPL opened the pull request.

DomiKoPL committed:
Add TT implementation.

DomiKoPL commented:
I have to add TT to Search. However search currently is just a simple function, so I would need to add it as a parameter. Weird. I have to think if I should do it correctly now or temporarily do it like that.

Gustav2137 commented:
TT should be either a global variable or we can wrap search into some class and put TT as a field there.

DomiKoPL commented:
Yes, we should. There should be a class that implements Engine like MinMaxEngine or sth.

Gustav2137 commented:
Oh that's what the engine class is for. I can refactor code to fit this template then.

Gustav2137 commented:
I have just committed changes we were discussing to #25. Feel free to review them.

Gustav2137 commented:
> Reference

Ok done.

DomiKoPL committed:
Add Lock-less TT implementation.

Implementation based on https://www.chessprogramming.org/Shared_Hash_Table\#Lockless.
TTData stores score, move, depth, age and type.

DomiKoPL committed:
Add TODO's for potential overflows.

DomiKoPL committed:
Change std::vector to std::unique_ptr.

std::vector requires elements to be copy-constructible, which std::atomic doesn't allow. I guess we won't be resizing the TT after the creation, so simple dynamic array should be enough for now.

DomiKoPL committed:
Add GoogleTest and test directory.

DomiKoPL committed:
Add TT library.

DomiKoPL committed:
Fix GoogleTest cmake.

DomiKoPL committed:
Add first TT test.

Gustav2137 commented on cpp_bestia/src/tt.cpp:
```diff
@@ -0,0 +1,83 @@
+
+uint8_t TTData::depth() const { return (data_ >> DEPTH_SHIFT) & DEPTH_MASK; }
+
+uint8_t TTData::age() const { return (data_ >> DEPTH_SHIFT) & DEPTH_MASK; }
```
Should be AGE_SHIFT and AGE_MASK.

Gustav2137 committed:
fix typo in tt.cpp

Gustav2137 commented:
lgtm, merging
